types.go

```diff
@@ -0,0 +1,55 @@
+package process
+
+import "context"
+
+// Injecter tags a struct as being an expected target of dependency injection.
+type Injecter interface {
+	// Inject populates the fields of the wrapped value in the given meta value.
+	Inject(ctx context.Context, meta *Meta) error
+}
+
+// Initializer wraps behavior that happens on application startup.
+type Initializer interface {
+	// Init is the hook invoked on application startup.
+	Init(ctx context.Context) error
+}
+
+// Process wraps behavior that happens continually through the application
+// lifecycle.
+type Process interface {
+	// Init is the hook invoked on application startup. The process hook is
+	// expected to be long-running. Returning early, prior to the given context
+	// being canceled, is generally an unexpected event.
+	Start(ctx context.Context) error
+}
+
+// Stoppable wraps a process with way to signal graceful exit.
+type Stoppable interface {
+	// Stop is the hookinvoked on a running process immediately prior to the
+	// root context being canceled. This method may exit immediately and is
+	// not expected to synchronize on Start returning.
+	Stop(ctx context.Context) error
+}
+
+// Finalizer wraps behavior that happens directly before application exit.
+type Finalizer interface {
+	// Finalize is the hook invoked directly before application exit.
+	Finalize(ctx context.Context) error
```

Comment:
Will this always be before application exit or can this be used to clean up any process that's ended, such as a process that can validly exit early?

Reply:
Finalizers are run after all processes have exited (cleanly or not). I feel like the situation you described can be cleanly implemented by a `defer` in the `Run`/`Start` method.